File: Backend/src/main/java/com/fairsplit/controller/UserController.java
package com.fairsplit.controller;

import com.fairsplit.model.User;
import com.fairsplit.service.UserService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Optional;

@RestController
@RequestMapping("/api/users")
public class UserController {
    
    @Autowired
    private UserService userService;
    
    @GetMapping
    public ResponseEntity<List<User>> getAllUsers() {
        List<User> users = userService.getAllUsers();
        return new ResponseEntity<>(users, HttpStatus.OK);
    }
    
    @GetMapping("/{id}")
    public ResponseEntity<User> getUserById(@PathVariable Long id) {
        Optional<User> user = userService.getUserById(id);
        return user.map(value -> new ResponseEntity<>(value, HttpStatus.OK))
                .orElseGet(() -> new ResponseEntity<>(HttpStatus.NOT_FOUND));
    }

     
    @GetMapping("/username/{username}")
    public ResponseEntity<User> getUserById(@PathVariable String username) {
        Optional<User> user = userService.getUserByUsername(username);
        return user.map(value -> new ResponseEntity<>(value, HttpStatus.OK))
                .orElseGet(() -> new ResponseEntity<>(HttpStatus.NOT_FOUND));
    }
    
    // create new user
    // catches illegal argument exception by service function, passes http BAD REQUEST response
    @PostMapping("/register")
    public ResponseEntity<User> createUser(@RequestBody User user) {
        try {
            User savedUser = userService.createUser(user);
            return new ResponseEntity<>(savedUser, HttpStatus.CREATED);
        } catch (IllegalArgumentException e) {
            return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
        }
    }

    
    // update user on id by putting a new User object to replace
    @PutMapping("/{id}")
    public ResponseEntity<User> updateUser(@PathVariable Long id, @RequestBody User userDetails) {
        try {
            Optional<User> updatedUser = userService.updateUser(id, userDetails);
            
            return updatedUser.map(user -> new ResponseEntity<>(user, HttpStatus.OK))
                    .orElseGet(() -> new ResponseEntity<>(HttpStatus.NOT_FOUND));
        } catch (IllegalArgumentException e) {
            // Handle case where username already exists
            return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
        }
    }

    // update user balance to a specific amount
    @PutMapping("/{id}/balance")
    public ResponseEntity<User> updateUserBalance(@PathVariable Long id, @RequestBody Double newAmount) {
        Optional<User> updatedUser = userService.updateUserBalance(id, newAmount);
        
        return updatedUser.map(user -> new ResponseEntity<>(user, HttpStatus.OK))
                .orElseGet(() -> new ResponseEntity<>(HttpStatus.NOT_FOUND));
    }

    @GetMapping("/group/{groupId}")
    public ResponseEntity<List<User>> getGroupUsers(@PathVariable Long groupId){
        List<User> users = userService.getFilteredUsers(groupId);
        return new ResponseEntity<>(users, HttpStatus.OK);
    }


    // add amount to user's current balance
    @PutMapping("/{id}/add-balance")
    public ResponseEntity<User> addUserAmount(@PathVariable Long id, @RequestBody Double amountToAdd) {
        Optional<User> updatedUser = userService.addUserAmount(id, amountToAdd);
        
        return updatedUser.map(user -> new ResponseEntity<>(user, HttpStatus.OK))
                .orElseGet(() -> new ResponseEntity<>(HttpStatus.NOT_FOUND));
    }
    
    // delete user by id
    @DeleteMapping("/{id}")
    public ResponseEntity<HttpStatus> deleteUser(@PathVariable Long id) {
        boolean deleted = userService.deleteUser(id);
        
        if (deleted) {
            return new ResponseEntity<>(HttpStatus.NO_CONTENT);
        } else {
            return new ResponseEntity<>(HttpStatus.NOT_FOUND);
        }
    }
}

File: Node-Backend/src/controllers/userController.ts
import { Request, Response } from 'express';
import { UserService } from '../services/userService.js';
import { AuthenticatedRequest } from '../middleware/auth.js';
import { CreateUserRequest, UpdateUserRequest } from '../types/user.js';

export class UserController {
  private userService: UserService;

  constructor() {
    this.userService = new UserService();
  }

  /**
   * Get all users
   */
  getAllUsers = async (req: Request, res: Response): Promise<void> => {
    try {
      const users = await this.userService.getAllUsers();
      res.status(200).json(users);
    } catch (error) {
      console.error('Error getting all users:', error);
      res.status(500).json({ error: 'Failed to fetch users' });
    }
  };

  /**
   * Get user by ID
   */
  getUserById = async (req: Request, res: Response): Promise<void> => {
    try {
      const { id } = req.params;
      if (!id) {
        res.status(400).json({ error: 'User ID is required' });
        return;
      }

      const user = await this.userService.getUserById(id);

      if (!user) {
        res.status(404).json({ error: 'User not found' });
        return;
      }

      res.status(200).json(user);
    } catch (error) {
      console.error('Error getting user by ID:', error);
      res.status(500).json({ error: 'Failed to fetch user' });
    }
  };

  /**
   * Get user by username (full user data - requires authentication)
   */
  getUserByUsername = async (req: Request, res: Response): Promise<void> => {
    try {
      const { username } = req.params;
      if (!username) {
        res.status(400).json({ error: 'Username is required' });
        return;
      }

      const user = await this.userService.getUserByUsername(username);

      if (!user) {
        res.status(404).json({ error: 'User not found' });
        return;
      }

      res.status(200).json(user);
    } catch (error) {
      console.error('Error getting user by username:', error);
      res.status(500).json({ error: 'Failed to fetch user' });
    }
  };

  /**
   * Get public user information by username (safe for public access)
   */
  getPublicUserByUsername = async (req: Request, res: Response): Promise<void> => {
    try {
      const { username } = req.params;
      if (!username) {
        res.status(400).json({ error: 'Username is required' });
        return;
      }

      const publicUser = await this.userService.getPublicUserByUsername(username);

      if (!publicUser) {
        res.status(404).json({ error: 'User not found' });
        return;
      }

      res.status(200).json(publicUser);
    } catch (error) {
      console.error('Error getting public user by username:', error);
      res.status(500).json({ error: 'Failed to fetch user' });
    }
  };

  /**
   * Create new user (requires authentication)
   */
  createUser = async (req: AuthenticatedRequest, res: Response): Promise<void> => {
    try {
      if (!req.user) {
        res.status(401).json({ error: 'Authentication required' });
        return;
      }

      const userData: CreateUserRequest = req.body;
      
      // Validate required fields
      if (!userData.username) {
        res.status(400).json({ error: 'Username is required' });
        return;
      }

      const user = await this.userService.createUser(userData, req.user.id);
      res.status(201).json(user);
    } catch (error) {
      console.error('Error creating user:', error);
      
      if (error instanceof Error && error.message === 'Username already exists') {
        res.status(400).json({ error: 'Username already exists' });
        return;
      }

      res.status(500).json({ error: 'Failed to create user' });
    }
  };

  /**
   * Update user
   */
  updateUser = async (req: AuthenticatedRequest, res: Response): Promise<void> => {
    try {
      const { id } = req.params;
      if (!id) {
        res.status(400).json({ error: 'User ID is required' });
        return;
      }

      const userDetails: UpdateUserRequest = req.body;

      // Users can only update their own profile
      if (req.user && req.user.id !== id) {
        res.status(403).json({ error: 'Can only update your own profile' });
        return;
      }

      const updatedUser = await this.userService.updateUser(id, userDetails);

      if (!updatedUser) {
        res.status(404).json({ error: 'User not found' });
        return;
      }

      res.status(200).json(updatedUser);
    } catch (error) {
      console.error('Error updating user:', error);
      
      if (error instanceof Error && error.message === 'Username already exists') {
        res.status(400).json({ error: 'Username already exists' });
        return;
      }

      res.status(500).json({ error: 'Failed to update user' });
    }
  };

  /**
   * Update user balance to a specific amount
   */
  updateUserBalance = async (req: Request, res: Response): Promise<void> => {
    try {
      const { id } = req.params;
      if (!id) {
        res.status(400).json({ error: 'User ID is required' });
        return;
      }

      const { amount } = req.body;

      if (typeof amount !== 'number') {
        res.status(400).json({ error: 'Amount must be a number' });
        return;
      }

      const updatedUser = await this.userService.updateUserBalance(id, amount);

      if (!updatedUser) {
        res.status(404).json({ error: 'User not found' });
        return;
      }

      res.status(200).json(updatedUser);
    } catch (error) {
      console.error('Error updating user balance:', error);
      res.status(500).json({ error: 'Failed to update user balance' });
    }
  };

  /**
   * Add amount to user's current balance
   */
  addUserAmount = async (req: Request, res: Response): Promise<void> => {
    try {
      const { id } = req.params;
      if (!id) {
        res.status(400).json({ error: 'User ID is required' });
        return;
      }

      const { amount } = req.body;

      if (typeof amount !== 'number') {
        res.status(400).json({ error: 'Amount must be a number' });
        return;
      }

      const updatedUser = await this.userService.addUserAmount(id, amount);

      if (!updatedUser) {
        res.status(404).json({ error: 'User not found' });
        return;
      }

      res.status(200).json(updatedUser);
    } catch (error) {
      console.error('Error adding user amount:', error);
      res.status(500).json({ error: 'Failed to add user amount' });
    }
  };

  /**
   * Get users by group
   */
  getGroupUsers = async (req: Request, res: Response): Promise<void> => {
    try {
      const { groupId } = req.params;
      if (!groupId) {
        res.status(400).json({ error: 'Group ID is required' });
        return;
      }

      const users = await this.userService.getFilteredUsers(groupId);
      res.status(200).json(users);
    } catch (error) {
      console.error('Error getting group users:', error);
      res.status(500).json({ error: 'Failed to fetch group users' });
    }
  };

  /**
   * Delete user (soft delete)
   */
  deleteUser = async (req: AuthenticatedRequest, res: Response): Promise<void> => {
    try {
      const { id } = req.params;
      if (!id) {
        res.status(400).json({ error: 'User ID is required' });
        return;
      }

      // Users can only delete their own account
      if (req.user && req.user.id !== id) {
        res.status(403).json({ error: 'Can only delete your own account' });
        return;
      }

      const deleted = await this.userService.deleteUser(id);

      if (!deleted) {
        res.status(404).json({ error: 'User not found' });
        return;
      }

      res.status(204).send();
    } catch (error) {
      console.error('Error deleting user:', error);
      res.status(500).json({ error: 'Failed to delete user' });
    }
  };

  /**
   * Get current user profile
   */
  getCurrentUser = async (req: AuthenticatedRequest, res: Response): Promise<void> => {
    try {
      if (!req.user) {
        res.status(401).json({ error: 'Authentication required' });
        return;
      }

      const user = await this.userService.getUserById(req.user.id);

      if (!user) {
        res.status(404).json({ error: 'User profile not found' });
        return;
      }

      res.status(200).json(user);
    } catch (error) {
      console.error('Error getting current user:', error);
      res.status(500).json({ error: 'Failed to fetch user profile' });
    }
  };
} 